### ISSUE TITLE
Failed to create ResourceBinding for resources with `::` in the name

### ISSUE THREAD

**Issue by a7i** [karmada-io/karmada#4681]
**What happened**:

If a Role has two colons in the name `::`, then a ResourceBinding nor Work is created. Karmada replaced colon with dot (not sure why), but this causes issues as the name ends up having two dots in it which is not RFC 1123 compliant

```
Apply cluster policy(duplicate) failed: ResourceBinding.work.karmada.io "system..leader-locking-kube-controller-manager-role" is invalid: metadata.name: Invalid value: "system..leader-locking-kube-controller-manager-role": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
```

**What you expected to happen**:
The name to be sanitized by karmada-controller-manager

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Karmada version: v1.9.0
- kubectl-karmada or karmadactl version (the result of `kubectl-karmada version` or `karmadactl version`):
- Others:

**Comment by XiShanYongYe-Chang** [comment]
@a7i Thanks a lot for your feedback!

> To be fair, we should have never propagated this role (since it's a system role)

I agree with you that propagate system Role don't seem to have a reasonable usage scenario. We should consider banning propagation [system Role](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#referring-to-subjects) (and other RBAC resources, including RoleBinding, ClusterRole, and ClusterRoleBinding).

By the way, how did you discover the problem? Are you trying to propagate a system Role?

**Comment by a7i** [comment]
@XiShanYongYe-Chang that is correct, I accidentally propagated a system role.

With that being said, `::` is valid in names for clusterrole/clusterolebinding/roles/rolebindings. See example below:

```
$ kubectl create role "foo::bar" --verb=get --resource=pods

role.rbac.authorization.k8s.io/foo::bar created
```

```
$ kubectl get role "foo::bar" -oyaml

apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: foo::bar
  namespace: default
rules:
- apiGroups:
  - ""
  resources:
  - pods
  verbs:
  - get
```

**Comment by XiShanYongYe-Chang** [comment]
Ask @RainbowMango to help take a look.

**Comment by RainbowMango** [comment]
> Karmada replaces colon with dot (not sure why), 

That is because the CRD resources can not accept names with `:`, the comments here explained the reason.
https://github.com/karmada-io/karmada/blob/e7300c3c1850629463af95f54f31a032a1890ce1/pkg/util/names/names.go#L65-L71

This is indeed a tradeoff to do the replacement. :(

> To be fair, we should have never propagated this role (since it's a system role), but regardless it should be sanitized.

The trouble is that we can not assume there is one or two colons in the name. There may be more. 

Echo from https://kubernetes.io/docs/reference/access-authn-authz/rbac/#referring-to-subjects:
> Caution: The prefix system: is reserved for Kubernetes system use, so you should ensure that you don't have users or groups with names that start with system: by accident. Other than this special prefix, the RBAC authorization system does not require any format for usernames.

Given the RBAC system doesn't require any format, that means the users might create resources with any number of colons, like:
```bash
# kubectl create role "foo:::bar" --verb=get --resource=pods  // 3 colons
role.rbac.authorization.k8s.io/foo:::bar created
# kubectl create role "foo::::bar" --verb=get --resource=pods // 4 colons
role.rbac.authorization.k8s.io/foo::::bar created
# # kubectl create role "foo:::::bar" --verb=get --resource=pods // 5 colons
role.rbac.authorization.k8s.io/foo:::::bar created
```
How do we handle these cases? Or shall we handle that?

**Comment by a7i** [comment]
Perhaps we could replace `:` with `-` instead of `.`

Given that `fo----bar` is still a valid domain

**Comment by a7i** [comment]
```
kubectl create role "foo:::bar" --verb=get --resource=pods

role.rbac.authorization.k8s.io/foo:::bar created
```

```
kubectl label role "foo:::bar" foo---bar=test
 
role.rbac.authorization.k8s.io/foo:::bar labeled
```

```
kubectl get role foo:::bar -L foo---bar
NAME        CREATED AT             FOO---BAR
foo:::bar   2024-03-14T03:05:57Z   test
```

**Comment by XiShanYongYe-Chang** [comment]
This is a feasible method.

**Comment by RainbowMango** [comment]
> Perhaps we could replace : with - instead of .

Yeah, but it might be a breaking change incompatible with previous versions. That is concerning. right?

**Comment by XiShanYongYe-Chang** [comment]
It seems that because this solution is a break change, we are consistently blocking.

Do you have any more ideas?
@a7i @RainbowMango

**Comment by XiShanYongYe-Chang** [comment]
In the case of two colons, it's a new beginning. In the case of one colon? Like this:
`foo:bar` -> `foo.bar`

**Comment by a7i** [comment]
What are your thoughts on this? https://github.com/karmada-io/karmada/issues/4681#issuecomment-1996306581

For single colon, leave existing functionality to use a dot, but when it's two, use a dash. 

`foo:bar` -> `foo.bar`
`foo::bar` -> `foo--bar`
`foo:::bar` -> `foo---bar`

**Comment by XiShanYongYe-Chang** [comment]
It works, but it looks a little weird, how do you think @RainbowMango @chaunceyjiang?
/cc @RainbowMango @chaunceyjiang

**Comment by a7i** [comment]
I agree, it's not very elegant.
